#59249 - Add qgis.env.in and cmake step to generate when using vcpkg

NathanW2 opened the pull request:
## Description

Adds a new step to export `qgis.env` into the output build folder for vcpkg based Windows builds

Part fixes for #59249

Some notes:

- I put the qgis.env.in inside a new `ms-windows/dev` folder. Might be a better home for it but wasn't sure
- I haven't done cmake in a long while so this might not be 100% correct place to put this call but tried to align it with other build output logic
- I use `VCPKG_APPLOCAL_DEPS=OFF` haven't tested with it turned on and if that changes any paths (will do another clean build to confirm I have this right)
- Suspect I haven't covered off all the paths required for everything QGIS e.g GRASS.  This is only a start to make sure it runs and loads Python.

Plan to use the `ms-windows/dev` folder for other Windows based local developer scripts into here for new Windows developers. Bootstrap environment scripts, std local development cmake args etc and powershell scripts to install cmake, VS etc using `winget` (This will be in a different PR when I get some time to clean it up)

# TODO 

- [x] Fix paths to `C:\Windows\system32` as placeholders
- [x] Confirm clean build will load correctly with only given vars

NathanW2 committed:
Add qgis.env.in and cmake step to generate when using vcpkg

nyalldawson commented:
@m-kuhn look ok to you?

NathanW2 committed:
Fix paths for windows system32 folders

m-kuhn commented on CMakeLists.txt:
```diff
@@ -1102,6 +1103,14 @@ set (CMAKE_LIBRARY_OUTPUT_DIRECTORY ${QGIS_OUTPUT_DIRECTORY}/${QGIS_LIB_SUBDIR})
+if(WITH_VCPKG)
+  if(WIN32)
+    set(VCPKG_TARGET_PATH "${VCPKG_INSTALL_PREFIX}/${VCPKG_TARGET_TRIPLET}")
+    configure_file(${CMAKE_SOURCE_DIR}/ms-windows/dev/qgis.env.in ${QGIS_OUTPUT_DIRECTORY}/${QGIS_BIN_SUBDIR}/qgis.env)
```
```suggestion
    configure_file("${CMAKE_SOURCE_DIR}/ms-windows/dev/qgis.env.in" "${QGIS_OUTPUT_DIRECTORY}/${QGIS_BIN_SUBDIR}/qgis.env")
```

Is more robust in case of spaces in the path

m-kuhn commented:
What a pleasant surprise to see you here @NathanW2 
I like the approach

NathanW2 committed:
Quote qgis.env write out paths in cmake

NathanW2 commented:
> What a pleasant surprise to see you here @NathanW2 I like the approach

😃 I am hoping to get back into a little bit of QGIS dev here and there.   Would like to improve our Windows QGIS setup process for new developers etc so this is just a good start to get it on track.

Have updated the cmake file to quote the output.

NathanW2 commented:
At the moment I still get the following error loading:

```
"Authentication system is DISABLED:\nQCA's OpenSSL plugin (qca-ossl) is missing"
```

But unsure if that is mine or a general problem in the Qt6 builds.  Have added `SDK\bin\Qca\crypto`  into PATH in .env but doesn't work yet.  WIll debug some more and update the env once I know.

m-kuhn commented:
It looks for the qca plugins next to the qt plugins (some subfolder, check the build artifacts for the specific path).

NathanW2 committed:
Fix path to Qca plugins

NathanW2 commented:
This .env won't work with bundled vcpkg packages though at the moment the way I have it because those paths are different (maybe I will need to confirm and test) my goal at the moment is to just get it to work for local builds without having to copy all the SDK stuff into the build folders when just doing local development.

NathanW2 commented:
@m-kuhn @nyalldawson I have put this  back to draft just while I was checking if there was anything else I needed to do.  

The main goal was to not have to copy the SDK libraries out into the build folder when doing local development for 1) speed up build time 2) save disk space when I have all the binaries already in a different place.

I'm happy to take what ever approch makes sense here.    I checked the last Qt6 github from here and it seems it still loads fine and doesn't include the qgis.env file so it doesn't break that process.

If it was ever included it would need to use relative paths and not absolute like it is here.

m-kuhn commented:
> The main goal was to not have to copy the SDK libraries out into the build folder when doing local development for 1) speed up build time 2) save disk space when I have all the binaries already in a different place.

I do fully agree with this

> I'm happy to take what ever approch makes sense here. I checked the last Qt6 github from here and it seems it still loads fine and doesn't include the qgis.env file so it doesn't break that process.

What exactly are you referring to here? That the artifacts from the github pipeline of this pull request still works as expected?

> If it was ever included it would need to use relative paths and not absolute like it is here.

Relative to what?
I think having an absolute path to the SDK makes sense, this is a build machine local configuration.

Is the current approach for Qca working as expected?

NathanW2 commented:
> > The main goal was to not have to copy the SDK libraries out into the build folder when doing local development for 1) speed up build time 2) save disk space when I have all the binaries already in a different place.
> 
> I do fully agree with this
> 
> > I'm happy to take what ever approch makes sense here. I checked the last Qt6 github from here and it seems it still loads fine and doesn't include the qgis.env file so it doesn't break that process.
> 
> What exactly are you referring to here? That the artifacts from the github pipeline of this pull request still works as expected?

Yeah making sure the current build artifacts from this branch don't include the file and still load correctly without it.

> > If it was ever included it would need to use relative paths and not absolute like it is here.
> 
> Relative to what? I think having an absolute path to the SDK makes sense, this is a build machine local configuration.

Ok was just checking this is fine for this approach.  I assumed as much given it's really for ease of local development and not packaging 

> Is the current approach for Qca working as expected?

Yep it is working once I put the Qca path as part of QT_PLUGIN_PATH

I have a few more fixes and I will leave this open for day for any more feedback.

NathanW2 committed:
Remove QGIS_PREFIX_PATH from qgis.env and update cmake rules

QGIS_PREFIX_PATH isn't required as qgisbuildpath.txt is in output
and is used at started to set those paths

NathanW2 commented:
@m-kuhn Ok I think this is good now.  Happy for you to merge when ready or I can.

m-kuhn commented on ms-windows/dev/qgis.env.in:
```diff
@@ -0,0 +1,5 @@
+PATH=@VCPKG_TARGET_PATH@\bin;$ENV{WINDIR};$ENV{WINDIR}\system32\WBem
```
What resolves the `$ENV{}` parts in here? If they are meant for cmake configure_file, should they be with `@` or are they for runtime?

NathanW2 replied:
Seems it's a supported cmake replacement and can be used to look up other environment variables. If you set `@ONLY` it will turn that feature off.

> [Variables](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#cmake-language-variables) referenced in the input file content as @VAR@, ${VAR}, $CACHE{VAR}, and [environment variables](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#cmake-language-environment-variables) referenced as $ENV{VAR}, will each be replaced with the current value of the variable, or the empty string if the variable is not defined

So this resolves the `windir` environment variable for system path e.g `C:\Windows32`

This is only needed because we are setting the full `PATH` environonment in our `main.cpp` of QGIS else we really could leave all this off.

m-kuhn replied:
Ah ok, I've never mixed the two styles

NathanW2 replied:
@m-kuhn yeah I thought it made sense here only so that I didn't define a cmake variable only to use it in once place and not clear where.

nicogodet commented:
Hi @NathanW2, `qgis.env` file is not present in build artifact from CI.

<img width="632" height="330" alt="image" src="https://github.com/user-attachments/assets/c078d30d-2b48-40dc-8312-fdf6c51f2fda" />

So it breaks some functionality (see https://github.com/qgis/QGIS/issues/63221).

Should we extend variable list to match the one in OSGeo4W package?

```
PATH=C:\PROGRA~1\QGIS\LTR\apps\qgis-ltr\bin;C:\PROGRA~1\QGIS\LTR\apps\grass\grass84\lib;C:\PROGRA~1\QGIS\LTR\apps\grass\grass84\bin;C:\PROGRA~1\QGIS\LTR\apps\qt5\bin;C:\PROGRA~1\QGIS\LTR\apps\Python312\Scripts;C:\PROGRA~1\QGIS\LTR\bin;C:\Windows\system32;C:\Windows;C:\Windows\system32\WBem
GDAL_DATA=C:\PROGRA~1\QGIS\LTR\apps\gdal\share\gdal
GDAL_DRIVER_PATH=C:\PROGRA~1\QGIS\LTR\apps\gdal\lib\gdalplugins
GDAL_FILENAME_IS_UTF8=YES
PDAL_DRIVER_PATH=C:\PROGRA~1\QGIS\LTR\apps\pdal\plugins
GISBASE=C:\PROGRA~1\QGIS\LTR\apps\grass\grass84
GRASS_PROJSHARE=C:\PROGRA~1\QGIS\LTR\share\proj
GRASS_PYTHON=C:\PROGRA~1\QGIS\LTR\bin\python3.exe
GRASS_COMPATIBILITY_TEST=0
OSGEO4W_ROOT=C:\PROGRA~1\QGIS\LTR
PROJ_DATA=C:\PROGRA~1\QGIS\LTR\share\proj
PYTHONHOME=C:\PROGRA~1\QGIS\LTR\apps\Python312
PYTHONPATH=C:\PROGRA~1\QGIS\LTR\apps\grass\grass84\etc\python;
PYTHONUTF8=1
QGIS_PREFIX_PATH=C:/PROGRA~1/QGIS/LTR/apps/qgis-ltr
QT_PLUGIN_PATH=C:\PROGRA~1\QGIS\LTR\apps\qgis-ltr\qtplugins;C:\PROGRA~1\QGIS\LTR\apps\qt5\plugins
VSI_CACHE=TRUE
VSI_CACHE_SIZE=1000000
O4W_QT_PREFIX=C:/PROGRA~1/QGIS/LTR/apps/Qt5
O4W_QT_BINARIES=C:/PROGRA~1/QGIS/LTR/apps/Qt5/bin
O4W_QT_PLUGINS=C:/PROGRA~1/QGIS/LTR/apps/Qt5/plugins
O4W_QT_LIBRARIES=C:/PROGRA~1/QGIS/LTR/apps/Qt5/lib
O4W_QT_TRANSLATIONS=C:/PROGRA~1/QGIS/LTR/apps/Qt5/translations
O4W_QT_HEADERS=C:/PROGRA~1/QGIS/LTR/apps/Qt5/include
QGIS_WIN_APP_NAME=QGIS 3.40\QGIS Desktop 3.40.9
SSL_CERT_DIR=C:\PROGRA~1\QGIS\LTR\apps\openssl\certs
SSL_CERT_FILE=C:\PROGRA~1\QGIS\LTR\bin\curl-ca-bundle.crt
```

nicogodet commented:
I hit this "issue" again, I think we should document `VCPKG_APPLOCAL_DEPS` should be OFF in install.md. Any thoughts @m-kuhn ?
